File: src/modules/organization/services/organization-delete.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { PrismaService } from 'src/infra/database/prisma.service';
import { UserRequest } from 'src/shared/types/user-request';
import { userOrganizationValidator } from 'src/shared/validators/user-organization.validator';
import { OrganizationEntity } from '../domain/entities/organization.entity';
import { organizationDeleteValidator } from '../validators/organization-delete.validator';
import { OrganizationDeleteAvatarService } from './organization-delete-avatar.service';

@Injectable()
export class OrganizationDeleteService {
  constructor(
    private prisma: PrismaService,
    private organizationDeleteAvatarService: OrganizationDeleteAvatarService,
  ) {}
  async execute(
    id: number,
    userRequest: UserRequest,
  ): Promise<OrganizationEntity> {
    const organization = await this.prisma.organization.findFirst({
      where: {
        id: id,
        deletedAt: null,
      },
      include: {
        avatar: true,
      },
    });

    if (!organization)
      throw new NotFoundException('Organização não encontrada!');

    await userOrganizationValidator(id, userRequest);

    await organizationDeleteValidator(id);

    if (organization.avatar)
      await this.organizationDeleteAvatarService.execute(id, userRequest);

    const updatedOrganization = await this.prisma.organization.update({
      data: {
        deletedAt: new Date(),
      },
      where: {
        id: id,
      },
    });

    return new OrganizationEntity(updatedOrganization);
  }
}

File: src/modules/organization/services/organization_delete_service.py
from django.utils import timezone
from src.shared.types.user_request import UserRequest
from src.modules.organization.models import Organization
from src.shared.exceptions.not_found_exception import NotFoundException
from src.shared.validators.user_organization_validator import (
    user_organization_validator,
)
from src.modules.organization.validators.organization_delete_validator import (
    organization_delete_validator,
)
from src.modules.organization.services.organization_delete_avatar_service import (
    OrganizationDeleteAvatarService,
)


class OrganizationDeleteService:
    @classmethod
    def execute(self, id: int, user_request: UserRequest):

        organization = Organization.objects.filter(id=id, deleted_at=None).first()

        if organization is None:
            raise NotFoundException("Organização não encontrada!")

        user_organization_validator(id, user_request)
        organization_delete_validator(id)

        if organization.avatar is not None:
            organization = OrganizationDeleteAvatarService.execute(id, user_request)

        organization.deleted_at = timezone.now()
        organization.save()

        return organization
